File: android/src/main/java/fr/wavyapp/reactNativeStarPrinter/PrintCommands.kt
package fr.wavyapp.reactNativeStarPrinter

import com.starmicronics.stario10.starxpandcommand.printer.Alignment
import com.starmicronics.stario10.starxpandcommand.printer.CutType

sealed interface PrinterCommand

enum class Action: PrinterCommand {
  CUT,
  PARTIAL_CUT,
  FULL_DIRECT,
  PARTIAL_DIRECT,
  PAPER_FEED,
  FEED_LINE,
  PRINT_RULED_LINE;

  companion object {
    @Throws(IllegalArgumentException::class)
    fun fromString(value: String): Action {
      return when (value) {
        "cut" -> CUT
        "partial-cut" -> PARTIAL_CUT
        "full-direct" -> FULL_DIRECT
        "partial-direct" -> PARTIAL_DIRECT
        "paper-feed" -> PAPER_FEED
        "feed-line" -> FEED_LINE
        "print-line-separator" -> PRINT_RULED_LINE
        else -> throw IllegalArgumentException("$value is not a valid printer action")
      }
    }

    @Throws(IllegalArgumentException::class)
    fun toCutType(action: Action): CutType {
      return when (action) {
        CUT -> CutType.Full
        PARTIAL_CUT -> CutType.Partial
        FULL_DIRECT -> CutType.FullDirect
        PARTIAL_DIRECT -> CutType.PartialDirect
        else -> throw IllegalArgumentException("Could not tie $action to a valid CutType")
      }
    }
  }
}

enum class PrintDataType {
  TEXT,
  IMAGE,
  BARCODE;

  companion object {
    @Throws(IllegalArgumentException::class)
    fun fromString(value: String): PrintDataType {
      return when (value) {
        "text" -> TEXT
        "image" -> IMAGE
        "barcode" -> BARCODE
        else -> throw IllegalArgumentException("$value is not a valid print data type")
      }
    }
  }
}

@Throws(IllegalArgumentException::class)
fun alignmentFromString(value: String): Alignment {
  return when (value) {
    "center" -> Alignment.Center
    "left" -> Alignment.Left
    "right" -> Alignment.Right
    else -> throw IllegalArgumentException("Invalid align value $value")
  }
}

data class Style(
  val align: Alignment?,
  val barWidth: Int?,
  val bold: Boolean?,
  val diffusion: Boolean?,
  val threshold: Int?,
  val width: Int?,
  val height: Int?,
  val heightExpansion: Int?,
  val widthExpansion: Int?,
  val underlined: Boolean?
) {
  companion object {
    @Throws(IllegalArgumentException::class)
    fun fromMap(style: Map<*, *>): Style {
      style["align"]?.let {alignValue ->
        if (alignValue !is String) {
          throw IllegalArgumentException("Invalid align value type ${alignValue::class.simpleName}")
        }
      }
      style["barWidth"]?.let {barWidthValue ->
        // As per documentation React Native will convert integers to double https://reactnative.dev/docs/native-modules-android#argument-types
        if (barWidthValue !is Double) {
          throw IllegalArgumentException("Invalid barWidth value type ${barWidthValue::class.simpleName}")
        }
      }
      style["bold"]?.let {boldValue ->
        if (boldValue !is Boolean) {
          throw IllegalArgumentException("Invalid bold value type ${boldValue::class.simpleName}")
        }
      }
      style["diffusion"]?.let {diffusionValue ->
        if (diffusionValue !is Boolean) {
          throw IllegalArgumentException("Invalid diffusion value type ${diffusionValue::class.simpleName}")
        }
      }
      style["threshold"]?.let {thresholdValue ->
        // As per documentation React Native will convert Int to double https://reactnative.dev/docs/native-modules-android#argument-types
        if (thresholdValue !is Double) {
          throw IllegalArgumentException("Invalid threshold value type ${thresholdValue::class.simpleName}")
        }
      }
      style["width"]?.let {widthValue ->
        // As per documentation React Native will convert Int to double https://reactnative.dev/docs/native-modules-android#argument-types
        if (widthValue !is Double) {
          throw IllegalArgumentException("Invalid width value type ${widthValue::class.simpleName}")
        }
      }
      style["height"]?.let {heightValue ->
        // As per documentation React Native will convert Int to double https://reactnative.dev/docs/native-modules-android#argument-types
        if (heightValue !is Double) {
          throw IllegalArgumentException("Invalid height value type ${heightValue::class.simpleName}")
        }
      }
      style["heightExpansion"]?.let {heightExpansionValue ->
        // As per documentation React Native will convert Int to double https://reactnative.dev/docs/native-modules-android#argument-types
        if (heightExpansionValue !is Double) {
          throw IllegalArgumentException("Invalid heightExpansion value type ${heightExpansionValue::class.simpleName}")
        }
      }
      style["widthExpansion"]?.let {widthExpansionValue ->
        // As per documentation React Native will convert Int to Double https://reactnative.dev/docs/native-modules-android#argument-types
        if (widthExpansionValue !is Double) {
          throw IllegalArgumentException("Invalid widthExpansion value type ${widthExpansionValue::class.simpleName}")
        }
      }
      style["underlined"]?.let {underlinedValue ->
        if (underlinedValue !is Boolean) {
          throw IllegalArgumentException("Invalid underlined value type ${underlinedValue::class.simpleName}")
        }
      }

      return Style(
          (style["align"] as? String)?.let{ alignValue -> alignmentFromString(alignValue) },
          (style["barWidth"] as? Double)?.toInt(),
          style["bold"] as Boolean?,
          style["diffusion"] as Boolean?,
          (style["threshold"] as? Double)?.toInt(),
          (style["width"] as? Double)?.toInt(),
          (style["height"] as? Double)?.toInt(),
          (style["heightExpansion"] as? Double)?.toInt(),
          (style["widthExpansion"] as? Double)?.toInt(),
          style["underlined"] as Boolean?,
        )
    }
  }
}

data class Print(val data: String, val type: PrintDataType, val style: Style? = null): PrinterCommand
data class PrinterAction(val action: Action, val args: Map<*, *>?): PrinterCommand

fun printerCommandFactory(command: Map<String, Any?>): PrinterCommand {
  (command["data"] as? String)?.let { data ->
    return Print(
      data,
      (command["type"] as? String)?.let { type ->  PrintDataType.fromString(type) } ?: PrintDataType.TEXT,
      (command["style"] as? Map<*, *>)?.let { style -> Style.fromMap(style) }
    )
  }
  (command["action"] as? String)?.let { action ->
    return PrinterAction(Action.fromString(action), command["actionArguments"] as? Map<*,*>)
  }
  throw IllegalArgumentException("The command $command has no data nor action")
}

File: ios/ReactNativeStarPrinter/PrintCommands.swift
enum Action: String {
  case cut = "cut"
  case partialCut = "partial-cut"
  case fullDirect = "full-direct"
  case partialDirect = "partial-direct"
  case paperFeed = "paper-feed"
  case feedLine = "feed-line"
  case printRuledLine = "print-line-separator"
  
  static func fromString(value: String) throws -> Action {
    return switch value {
      case "cut":
          .cut
      case "partial-cut":
          .partialCut
      case "full-direct":
          .fullDirect
      case "partial-direct":
          .partialDirect
      case "paper-feed":
          .paperFeed
      case "feed-line":
          .feedLine
      case "print-line-separator":
          .printRuledLine
      default:
        throw ReactNativeStarPrinterError.invalidArgument(message: "\(value) is not a valid printer action")
    }
  }
}

enum PrintDataType: String {
  case text = "text"
  case image = "image"
  case barcode = "barcode"
  
  static func fromString(value: String) throws -> PrintDataType {
    return switch value {
      case "text":
          .text
      case "image":
          .image
      case "barcode":
          .barcode
      default:
        throw ReactNativeStarPrinterError.invalidArgument(message: "\(value) is not a valid print data type")
    }
  }
}

enum Align: String {
  case center = "center"
  case left = "left"
  case right = "right"
  
  static func fromString(value: String) throws -> Align {
    return switch value {
      case "center":
          .center
      case "left":
          .left
      case "right":
          .right
      default:
        throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid align value \(value)")
    }
  }
}

struct Style {
  let align: Align?
  let barWidth, height, heightExpansion, threshold, width, widthExpansion: Int?
  let bold, diffusion, underlined: Bool?
  
  init(
    align: Align?,
    barWidth: Int?,
    bold: Bool?,
    diffusion: Bool?,
    height: Int?,
    heightExpansion: Int?,
    threshold: Int?,
    underlined: Bool?,
    width: Int?,
    widthExpansion: Int?
  ) {
    self.align = align
    self.barWidth = barWidth
    self.bold = bold
    self.diffusion = diffusion
    self.height = height
    self.heightExpansion = heightExpansion
    self.threshold = threshold
    self.underlined = underlined
    self.width = width
    self.widthExpansion = widthExpansion
  }
  
  static func fromDict(style: [String: Any?]) throws -> Style {
    if (style["align"] as? String == nil && style["align"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid align value type \(style["align"])")
    }
    if (style["barWidth"] as? Int == nil && style["barWidth"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid barWidth value type \(style["barWidth"])")
    }
    if (style["bold"] as? Bool == nil && style["bold"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid bold value type \(style["bold"])")
    }
    if (style["heightExpansion"] as? Int == nil && style["heightExpansion"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid heightExpansion value type \(style["heightExpansion"])")
    }
    if (style["underlined"] as? Bool == nil && style["underlined"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid underlined value type \(style["underlined"])")
    }
    if (style["widthExpansion"] as? Int == nil && style["widthExpansion"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid widthExpansion value type \(style["widthExpansion"])")
    }
    if (style["width"] as? Int == nil && style["width"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid width value type \(style["width"])")
    }
    if (style["height"] as? Int == nil && style["height"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid height value type \(style["height"])")
    }
    if (style["threshold"] as? Int == nil && style["threshold"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid threshold value type \(style["threshold"])")
    }
    if (style["diffusion"] as? Bool == nil && style["diffusion"] != nil) {
      throw ReactNativeStarPrinterError.invalidArgument(message: "Invalid diffusion value type \(style["diffusion"])")
    }
    
    return Style(
      align: try style["align"].map { try Align.fromString(value: $0 as? String ?? "center") },
      barWidth: style["barWidth"] as? Int,
      bold: style["bold"] as? Bool,
      diffusion: style["diffusion"] as? Bool,
      height: style["height"] as? Int,
      heightExpansion: style["heightExpansion"] as? Int,
      threshold: style["threshold"] as? Int,
      underlined: style["underlined"] as? Bool,
      width: style["width"] as? Int,
      widthExpansion: style["widthExpansion"] as? Int
    )
  }
}

protocol PrinterCommand {}

struct Print: PrinterCommand {
  let data: String
  let style: Style?
  let type: PrintDataType
  
  init(data: String, type: PrintDataType, style: Style?) {
    self.data = data
    self.style = style
    self.type = type
  }
}

struct PrinterAction: PrinterCommand {
  let action: Action
  let args: [String: Any?]?
  
  init(_ action: Action, _ args: [String: Any?]?) {
    self.action = action
    self.args = args
  }
}

func printerCommandFactory(_ command: [String: Any?]) throws -> PrinterCommand {
  if let data = command["data"] as? String {
    let type = command["type"] as? String ?? "text"
    let style = command["style"] as? [String: Any?]
    return Print(
      data: data,
      type: try PrintDataType.fromString(value: type),
      style: try style.map { try Style.fromDict(style: $0) }
    )
  }
  if let action = command["action"] as? String {
    return PrinterAction(try Action.fromString(value: action), command["actionArguments"] as? [String: Any?])
  }
  throw ReactNativeStarPrinterError.invalidArgument(message: "The command $command has no data nor action")
}

